feat: improve code block feature with Expressive Code

Hasenpfote opened the pull request:
I noticed that Expressive Code was briefly mentioned in #319, so I tried replacing the existing code block implementation with it.
Please note that this is a large-scale change, and there are still open issues related to the previous implementation.
For now, I’ve kept the features and design minimal.

**Related open issues:** #275
**Related open PRs:** #208 #283

close #319

Kayxue commented:
Seems that language badge not worked? 🤔 🤔 🤔

Kayxue commented:
I revert changes from tutorial by other person and change the codes with file changes, but it doesn't worked.
The effect is like this.
![CleanShot 2025-03-04 at 23 50 38@2x](https://github.com/user-attachments/assets/d5c81e96-6742-4e8e-9b0f-2ea7bc69f983)

Hasenpfote commented:
I'm currently reviewing your repository. I've switched the light theme to rose-pine-dawn for better clarity. Everything seems to be working fine, but am I missing something?

![example](https://github.com/user-attachments/assets/f2e77bcc-3de1-405f-8360-4f8338911033)

Kayxue commented:
> I'm currently reviewing your repository. I've switched the light theme to rose-pine-dawn for better clarity. Everything seems to be working fine, but am I missing something?
> 
> ![example](https://private-user-images.githubusercontent.com/1502467/419278112-f2e77bcc-3de1-405f-8360-4f8338911033.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDExNDkzNDksIm5iZiI6MTc0MTE0OTA0OSwicGF0aCI6Ii8xNTAyNDY3LzQxOTI3ODExMi1mMmU3N2JjYy0zZGUxLTQwNWYtODM2MC00ZjgzMzg5MTEwMzMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDVUMDQzMDQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjU1MTY0OTNlNzQwMmY5YjBhMmNlODZlYzk3Y2JiZDk1ODVhMmRmNzA3NjQ4ZTIxMTQ1NThjNDA4ZGQ2Y2E2YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.WsfB1u8asSTdBeUVPrtBAnaeXZN8Mbx298JgfW2GwZQ)
> 
> One possibility is that the theme change might still be referencing cached data. Refreshing the page might resolve it.
> 
> Additionally, one noticeable difference between github-light and light-plus themes is the red line on the tabs. I believe the provided image is indeed from the light-plus theme.
> 
> ![example2](https://private-user-images.githubusercontent.com/1502467/419281854-7526cf31-c972-423b-9b81-91075aded0b3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDExNDkzNDksIm5iZiI6MTc0MTE0OTA0OSwicGF0aCI6Ii8xNTAyNDY3LzQxOTI4MTg1NC03NTI2Y2YzMS1jOTcyLTQyM2ItOWI4MS05MTA3NWFkZWQwYjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDVUMDQzMDQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTljNmI4NTExYTlkZGM1NDYzMzIzNTEwZTAxNWRiNjZiYTdkYzczZGJiYWUzNDczOWY4MmJhY2RlN2U2ODg3NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.D80ipdUM6BGtgPCJUQCU4xNMgYXlfyMn9YfRy_XprKo)
> 
> [Themes - Expressive Code](https://expressive-code.com/guides/themes/)

Sorry I mean the language badge part

Hasenpfote commented:
The conditions for the badge to appear are as follows:

```ts
.frame:not(.has-title):not(.is-terminal)
```

You can check this in `src/plugins/expressive-code/language-badge.ts`.

The reason for this is to avoid redundancy and clutter on the screen:

- If there is a title, the language can be identified
- Terminals have different frame displays and can be differentiated by their titles

I don't think this is the best approach, but if you prefer to have the badge always displayed, you can simply modify the CSS.

baka-gourd commented:
related: https://github.com/saicaca/fuwari/issues/275

didntchooseaname commented:
Hello, there is an issue with the long labels on their own lines from [https://expressive-code.com/key-features/text-markers/](https://expressive-code.com/key-features/text-markers/) renders like this in the project :

![image](https://github.com/user-attachments/assets/5fec660c-a4aa-45a2-8c13-985702556d39)

The sentence is not displayed :

Codeblock in md file:

```python {"test badge":2-5}
  import time
  
  # Quick, count to ten!
  for i in range(10):
      # (but not *too* quick)
      time.sleep(0.5)
      print i
```

With ```python {"test badge":2-5} at first.

Should render like that : 
![image](https://github.com/user-attachments/assets/f0aa1472-c731-4c08-bbb5-6f75308b9327)

Hasenpfote commented:
@didntchooseaname
Thank you for the report. I have confirmed the issue in my local (prod) environment as well. It may take some time to identify the cause.

Hasenpfote commented:
@didntchooseaname
After troubleshooting, I found that the behavior changes depending on whether astro-compress is enabled. The same issue occurs with the latest version as well. Could you check if disabling astro-compress resolves the issue on your end?

```js
// astro.config.mjs
...
export default defineConfig({
  ...
  integrations: [
    ...,
    icon({
      include: {
        "preprocess: vitePreprocess(),": ["*"],
        "fa6-brands": ["*"],
        "fa6-regular": ["*"],
        "fa6-solid": ["*"],
      },
    }),
    svelte(),
    sitemap(),
/*
    Compress({
      CSS: false,
      Image: false,
      Action: {
        Passed: async () => true, // https://github.com/PlayForm/Compress/issues/376
      },
    }),
*/
    expressiveCode({
```

didntchooseaname commented:
> @didntchooseaname After troubleshooting, I found that the behavior changes depending on whether astro-compress is enabled. The same issue occurs with the latest version as well. Could you check if disabling astro-compress resolves the issue on your end?
> 
> ```js
> // astro.config.mjs
> ...
> export default defineConfig({
>   ...
>   integrations: [
>     ...,
>     icon({
>       include: {
>         "preprocess: vitePreprocess(),": ["*"],
>         "fa6-brands": ["*"],
>         "fa6-regular": ["*"],
>         "fa6-solid": ["*"],
>       },
>     }),
>     svelte(),
>     sitemap(),
>     Compress({
>       CSS: false,
>       HTML: false, // Add this
>       Image: false,
>       Action: {
>         Passed: async () => true, // https://github.com/PlayForm/Compress/issues/376
>       },
>     }),
>     expressiveCode({
> ```
> 
> Currently testing a method to resolve the issue while keeping the compression enabled.

@Hasenpfote 

I confirm that astro-compress cause the issue there, here is the result without:
![image](https://github.com/user-attachments/assets/375bd2e9-3fa4-4271-a144-e2003484b962)

Hasenpfote commented:
I've fixed an issue where astro-compress interfered with Expressive Code. While it may not be a perfect solution, it effectively addresses the problem. Additionally, I’ve added a new document for visual review.

CuB3y0nd commented on src/styles/expressive-code.css:
Should add margin to code block.

```css
.expressive-code {
  @apply mb-6;
}
```

Hasenpfote replied:
Thank you for the suggestion. This feature is in the proposal phase, maintaining Expressive Code defaults.

saicaca commented:
Thank you for your contribution!
I am now merging this into a working branch.
